modules/build/src/main/scala/scala/build/preprocessing/ExtractedDirectives.scala

```diff
@@ -0,0 +1,107 @@
+      val settings = new Settings
+      settings.setAllowStartWithoutAt(true)
+      settings.setAllowRequire(false)
+      val context = new Context(reporter, settings)
+      new UsingDirectivesProcessor(context)
+    }
+    val all = processor.extract(contentChars, true, true).asScala
+
+    def byKind(kind: UsingDirectiveKind) = all.find(_.getKind == kind).get
+
+    def getDirectives(directives: UsingDirectives) =
+      directives.getAst() match {
+        case ud: UsingDefs =>
+          ud.getUsingDefs().asScala
+        case _ =>
+          Nil
+      }
+
+    val codeDirectives           = byKind(UsingDirectiveKind.Code)
+    val specialCommentDirectives = byKind(UsingDirectiveKind.SpecialComment)
+    val plainCommentDirectives   = byKind(UsingDirectiveKind.PlainComment)
+
+    def reportWarning(msg: String, values: Seq[UsingDef], before: Boolean = true): Unit =
+      values.foreach { v =>
+        val astPos = v.getPosition()
+        val (start, end) =
+          if (before) (0, astPos.getColumn())
+          else (astPos.getColumn(), astPos.getColumn() + v.getSyntax.getKeyword.size)
+        val position = Position.File(path, (astPos.getLine(), start), (astPos.getLine(), end))
+        logger.diagnostic(msg, positions = Seq(position))
+      }
+
+    val usedDirectives =
+      if (!codeDirectives.getFlattenedMap().isEmpty()) {
+        val msg =
+          s"This using directive is ignored. File contains directives outside comments and those has higher precedence."
+        reportWarning(
+          msg,
+          getDirectives(plainCommentDirectives) ++ getDirectives(specialCommentDirectives)
+        )
+        codeDirectives
+      }
+      else if (!specialCommentDirectives.getFlattenedMap().isEmpty()) {
+        val msg =
+          s"This using directive is ignored. $changeToSpecialCommentMsg"
+        reportWarning(msg, getDirectives(plainCommentDirectives))
+        specialCommentDirectives
+      }
+      else {
+        reportWarning(changeToSpecialCommentMsg, getDirectives(plainCommentDirectives))
+        plainCommentDirectives
+      }
+
+    // All using directives should use just `using` keyword, no @using or require
+    reportWarning(
+      "Deprecated using directive syntax, please use keyword `using`.",
+      getDirectives(usedDirectives).filter(_.getSyntax() != UsingDirectiveSyntax.Using),
+      before = false
+    )
+
+    val flattened = usedDirectives.getFlattenedMap.asScala.toSeq
+    val strictDirectives =
+      flattened.map { case (k, l) => StrictDirective(k.getPath.asScala.mkString("."), l.asScala) }
+
+    val offset =
+      if (usedDirectives.getKind() != UsingDirectiveKind.Code) 0
+      else usedDirectives.getCodeOffset()
+    if (supportedDirectives.contains(usedDirectives.getKind()))
+      Right(ExtractedDirectives(offset, strictDirectives))
+    else
+      Left(new DirectiveErrors(::(s"Unsupported using directive kind ${usedDirectives.getKind}", Nil)))
```

Comment:
Also, I've just realized that we do not provide a location where are the problematic cases. I think we should use something similar `UnusedDirectiveError`

Reply:
> The message may be confusing since in general keyword using directives are supported but not in the context of Java files and message should reflect that.

Agree.

The problem is that higher layers cannot refine error messages from lower layers as they are opaque in the current error design. Here we call same function `ExtractedDirectives.from` from two different context: scala and java. Each of these context should enrich the message. If only `DirectiveErrors` was a case class...


I can change that but I am afraid that it will makes this PR much bigger.